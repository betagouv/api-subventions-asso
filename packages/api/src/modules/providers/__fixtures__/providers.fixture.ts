import { ProviderEnum } from "../../../@enums/ProviderEnum";
import { FullGrantProvider } from "../../grant/@types/FullGrantProvider";
import PaymentProvider from "../../payments/@types/PaymentProvider";
import DemandesSubventionsProvider from "../../subventions/@types/DemandesSubventionsProvider";

const mockApplicationProvider = service => ({
    ...service,
    isDemandesSubventionsProvider: true,
    rawToApplication: jest.fn(),
    getDemandeSubvention: jest.fn().mockResolvedValue([]),
});

const mockFullGrantProvider = service => ({
    ...service,
    isFullGrantProvider: true,
    rawToGrant: jest.fn().mockResolvedValue([]),
});

const mockPaymentProvider = service => ({
    ...service,
    isPaymentProvider: true,
    rawToPayment: jest.fn(),
    getPayments: jest.fn().mockResolvedValue([]),
});

const mockGrantProvider = provider => ({
    provider,
    isGrantProvider: true,
    getRawGrants: jest.fn().mockResolvedValue([]),
});

export const fullGrantProvidersFixtures: FullGrantProvider<unknown>[] = [
    mockFullGrantProvider({
        meta: {
            id: "provider-fullgrant-1",
            name: "PROVIDER_FULLGRANT_1",
            description: "API PROVIDER",
            type: ProviderEnum.api,
        },
    }),
    mockFullGrantProvider({
        meta: {
            id: "provider-fullgrant-2",
            name: "PROVIDER_FULLGRANT_2",
            description: "RAW PROVIDER",
            type: ProviderEnum.raw,
        },
    }),
];

export const applicationProvidersFixtures: DemandesSubventionsProvider<unknown>[] = [
    mockApplicationProvider({
        meta: {
            id: "provider-application-1",
            name: "PROVIDER_APPLICATION_1",
            description: "API PROVIDER",
            type: ProviderEnum.api,
        },
    }),
    mockApplicationProvider({
        meta: {
            id: "provider-application-2",
            name: "PROVIDER_APPLICATION_2",
            description: "RAW PROVIDER",
            type: ProviderEnum.raw,
        },
    }),
];

export const paymentProvidersFixtures: PaymentProvider<unknown>[] = [
    mockPaymentProvider({
        meta: {
            id: "provider-payment-1",
            name: "PROVIDER_PAYMENT_1",
            description: "RAW PROVIDER",
            type: ProviderEnum.raw,
        },
    }),
    mockPaymentProvider({
        meta: {
            id: "provider-payment-2",
            name: "PROVIDER_PAYMENT_2",
            description: "RAW PROVIDER",
            type: ProviderEnum.raw,
        },
    }),
];

export const grantProvidersFixtures = [
    ...fullGrantProvidersFixtures,
    ...applicationProvidersFixtures,
    ...paymentProvidersFixtures,
].map(mockGrantProvider);
