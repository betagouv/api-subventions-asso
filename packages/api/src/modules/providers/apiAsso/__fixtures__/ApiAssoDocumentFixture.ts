import { StructureDacDocumentDto, StructureRnaDocumentDto } from "../dto/StructureDto";

export const ApiAssoDocumentFixture = {
    asso: {
        documents: {
            nbDocDac: 1,
            document_dac: [
                {
                    url: "/fake/url",
                    nom: "nom fake",
                    meta: {
                        type: "BPA",
                        id_siret: 50922194100000,
                        etat: "courant",
                    },
                    time_depot: "2021-06-18 12:02:53",
                },
            ] as StructureDacDocumentDto[],
            nbDocRna: 1,
            document_rna: [
                {
                    url: "/fake/url",
                    type: "PV",
                    sous_type: "PV",
                    id: "ididididid",
                    lib_sous_type: "PV",
                    time: 1622557171,
                    annee: 2021,
                },
            ] as StructureRnaDocumentDto[],
        },
    },
};
