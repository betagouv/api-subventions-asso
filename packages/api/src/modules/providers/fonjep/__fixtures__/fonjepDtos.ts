import { DefaultObject } from "../../../../@types";

export const TIER_DTO = {
    Code: "Code",
    RaisonSociale: "RaisonSociale",
    EstAssociation: "EstAssociation",
    EstCoFinanceurPostes: "EstCoFinanceurPostes",
    EstFinanceurPostes: "EstFinanceurPostes",
    SiretOuRidet: "SiretOuRidet",
    CodePostal: "CodePostal",
    Ville: "Ville",
    ContactEmail: "ContactEmail",
};

export const TIER_DTO_WITH_NULLS = { ...TIER_DTO, Ville: null };

export const POSTE_DTO = {
    Code: "Code",
    DispositifId: 1,
    PstStatutPosteLibelle: "PstStatutPosteLibelle",
    PstRaisonStatutLibelle: "PstRaisonStatutLibelle",
    FinanceurPrincipalCode: "FinanceurPrincipalCode",
    FinanceurAttributeurCode: "FinanceurAttributeurCode",
    AssociationBeneficiaireCode: "AssociationBeneficiaireCode",
    AssociationImplantationCode: "AssociationImplantationCode",
    Annee: 2022,
    MontantSubvention: 1000,
    DateFinTriennalite: 44926,
    PstTypePosteCode: "PstTypePosteCode",
    PleinTemps: "Oui",
    DoublementUniteCompte: "Non",
};

export const POSTE_DTO_WITH_NULLS = { ...POSTE_DTO, AssociationBeneficiaireCode: null };

export const VERSEMENT_DTO = {
    PosteCode: "Code",
    PeriodeDebut: 44896,
    PeriodeFin: 44835,
    DateVersement: 44926,
    MontantAPayer: 1000,
    MontantPaye: 1000,
};

export const TYPE_POSTE_DTO = {
    Code: "Code",
    Libelle: "Libelle",
};

export const TYPE_POSTE_DTO_WITH_NULLS = { ...TYPE_POSTE_DTO, Libelle: null };

export const DISPOSITIF_DTO = {
    Id: 1,
    Libelle: "Libelle",
    FinanceurCode: "FinanceurCode",
};

export const DISPOSITIF_DTO_WITH_NULLS = { ...DISPOSITIF_DTO, FinanceurCode: null };

export const VERSEMENT_DTO_WITH_NULLS = { ...VERSEMENT_DTO, MontantAPayer: null };

export const TIER_DTOS = [TIER_DTO, TIER_DTO_WITH_NULLS];

export const POSTE_DTOS = [POSTE_DTO, POSTE_DTO_WITH_NULLS];

export const VERSEMENT_DTOS = [VERSEMENT_DTO, VERSEMENT_DTO_WITH_NULLS];

export const TYPE_POSTE_DTOS = [TYPE_POSTE_DTO, TYPE_POSTE_DTO_WITH_NULLS];

export const DISPOSITIF_DTOS = [DISPOSITIF_DTO, DISPOSITIF_DTO_WITH_NULLS];
