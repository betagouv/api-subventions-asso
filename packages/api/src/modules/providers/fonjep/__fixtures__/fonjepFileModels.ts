export const DATA_WITH_HEADER = [
    [
        {
            Code: 1234,
            RaisonSociale: "CENTRE D'INFO ****",
            EstAssociation: 'Oui',
            EstCoFinanceurPostes: 'Non',
            EstFinanceurPostes: 'Non',
            SiretOuRidet: '00000000000000',
            CodePostal: '00000',
            Ville: 'Paris',
            ContactEmail: 'exemple@beta.gouv.fr'
        }
    ],
    [
        {
            Code: 'J10540',
            DispositifId: 3,
            PstStatutPosteLibelle: 'Attribué',
            PstRaisonStatutLibelle: 'Reconduction',
            FinanceurPrincipalCode: '1234',
            FinanceurAttributeurCode: 1234,
            AssociationBeneficiaireCode: 1234,
            AssociationImplantationCode: 1234,
            Annee: 2017,
            MontantSubvention: 6666,
            DateFinTriennalite: 43465,
            PstTypePosteCode: 'FONJEP',
            PleinTemps: 'Oui',
            DoublementUniteCompte: 'Non'
        }
    ],
    // TODO: add versements
    [],
    [
        { Code: 'DOUBLE', Libelle: 'UNITE DE COMPTE DOUBLEE' },
        { Code: 'FONJEP', Libelle: 'Poste FONJEP' },
        { Code: 'PSTEMP', Libelle: 'POSTES EMPOI FONJEP (100 %)' },
        { Code: 'FONBIS', Libelle: 'PROLONGATION PAS DE FRAIS' }
    ],
    [
        { ID: 1, Libelle: "FONJEP Jeunesse éducation populaire", FinanceurCode: 10004 },
        { ID: 2, Libelle: "FONJEP Cohésion sociale", FinanceurCode: 10005 },
        { ID: 3, Libelle: "FONJEP Politique de la ville", FinanceurCode: 10008 },
        { ID: 4, Libelle: "FONJEP Guid'Asso", FinanceurCode: 10009 },
        { ID: 5, Libelle: "FONJEP Éducation à la citoyenneté et à la solidarité internationale (ECSI)", FinanceurCode: 10010 },
        { ID: 6, Libelle: "FONJEP Culture", FinanceurCode: 10012 },
        { ID: 7, Libelle: "FONJEP Centre de ressour ces et d'information des bénévoles", FinanceurCode: 10016 },
        { ID: 8, Libelle: "FONJEP Jeunes", FinanceurCode: 10017 }
    ]
]


export const RAW_DATA = [
    [
        [
            'Code',
            'RaisonSociale',
            'EstAssociation',
            'EstCoFinanceurPostes',
            'EstFinanceurPostes',
            'SiretOuRidet',
            'CodePostal',
            'Ville',
            'ContactEmail'
        ],
        [
            1234,
            "CENTRE D'INFO ****",
            'Oui',
            'Non',
            'Non',
            '00000000000000',
            '00000',
            'Paris',
            'exemple@beta.gouv.fr'
        ]
    ],
    [
        [
            'Code',
            'DispositifId',
            'PstStatutPosteLibelle',
            'PstRaisonStatutLibelle',
            'FinanceurPrincipalCode',
            'FinanceurAttributeurCode',
            'AssociationBeneficiaireCode',
            'AssociationImplantationCode',
            'Annee',
            'MontantSubvention',
            'DateFinTriennalite',
            'PstTypePosteCode',
            'PleinTemps',
            'DoublementUniteCompte'
        ],
        [
            'J10540', 3, 'Attribué',
            'Reconduction', '1234',
            1234, 1234,
            1234, 2017,
            6666, 43465,
            'FONJEP', 'Oui',
            'Non'
        ]
    ],
    // TODO: add versements
    [],
    [
        ['Code', 'Libelle'],
        ['DOUBLE', 'UNITE DE COMPTE DOUBLEE'],
        ['FONJEP', 'Poste FONJEP'],
        ['PSTEMP', 'POSTES EMPOI FONJEP (100 %)'],
        ['FONBIS', 'PROLONGATION PAS DE FRAIS']
    ],
    [
        ['ID', 'Libelle', 'FinanceurCode'],
        [1, 'FONJEP Jeunesse éducation populaire', 10004],
        [2, 'FONJEP Cohésion sociale', 10005],
        [3, 'FONJEP Politique de la ville', 10008],
        [4, "FONJEP Guid'Asso", 10009],
        [5, 'FONJEP Éducation à la citoyenneté et à la solidarité internationale (ECSI)', 100010],
        [6, 'FONJEP Culture', 100012],
        [7, "FONJEP Centre de ressources et d'information des bénévoles", 100016],
        [8, 'FONJEP Jeunes', 100017],
    ]
]
