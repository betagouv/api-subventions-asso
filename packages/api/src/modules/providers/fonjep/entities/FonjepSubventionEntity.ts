import * as ParseHelper from "../../../../shared/helpers/ParserHelper";
import { Siret } from "@api-subventions-asso/dto";
import { ParserInfo, ParserPath, DefaultObject } from "../../../../@types";
import IFonjepIndexedInformations from "../@types/IFonjepIndexedInformations";
import { formatCP } from "../../../../shared/helpers/DataFormatHelper";

export default class FonjepSubventionEntity {
    public static indexedLegalInformationsPath: DefaultObject<ParserPath | ParserInfo> = {
        siret: {
            path: ["Association", "SiretOuRidet"],
            adapter: value => {
                if (!value) return value;

                return value.replace(/ /g, "");
            }
        },
        name: ["Association", "RaisonSociale"]
    };

    public static indexedProviderInformationsPath: DefaultObject<ParserPath | ParserInfo> = {
        code_poste: ["Code"],
        dispositif: ["Dispositif", "Libelle"],
        montant_paye: {
            path: ["MontantSubvention"],
            adapter: value => {
                if (!value) return 0;

                return !value.length ? parseFloat(value) : 0;
            }
        },
        status: ["PstStatutPosteLibelle"],
        raison: ["PstRaisonStatutLibelle"],
        service_instructeur: ["Financeur", "RaisonSociale"],
        annee_demande: ["Annee"],
        date_fin_triennale: {
            path: ["DateFinTriennalite"],
            adapter: value => {
                if (!value) return value;
                return ParseHelper.ExcelDateToJSDate(Number(value));
            }
        },
        updated_at: ["updated_at"],
        unique_id: ["id"],
        type_post: ["TypePoste", "Libelle"],
        ville: ["Association", "Ville"],
        code_postal: {
            path: ["Association", "CodePostal"],
            adapter: formatCP
        },
        contact: ["Association", "ContactEmail"],
        plein_temps: ["PleinTemps"]
    };

    constructor(
        public legalInformations: {
            siret: Siret;
            name: string;
        },
        public indexedInformations: IFonjepIndexedInformations,
        public data: unknown
    ) {}
}
