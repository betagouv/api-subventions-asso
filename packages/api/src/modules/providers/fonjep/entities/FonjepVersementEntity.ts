type FonjepVersementEntity = {
    posteCode: string;
    periodeDebut: Date | null;
    periodeFin: Date | null;
    dateVersement: Date | null;
    montantAPayer: number | null;
    montantPaye: number | null;
};

export type PayedFonjepVersementEntity = FonjepVersementEntity & {
    periodeDebut: Date;
    periodeFin: Date;
    dateVersement: Date;
    montantAPayer: number;
    montantPaye: number;
};

export type FullFonjepVersementEntity = PayedFonjepVersementEntity & { posteCode: string };

export default FonjepVersementEntity;
