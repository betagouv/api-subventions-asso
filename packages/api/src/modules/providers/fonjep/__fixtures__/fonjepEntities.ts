import DEFAULT_ASSOCIATION from "../../../../../tests/__fixtures__/association.fixture";
import FonjepDispositifEntity from "../entities/FonjepDispositifEntity";
import FonjepPosteEntity from "../entities/FonjepPosteEntity";
import FonjepTiersEntity from "../entities/FonjepTiersEntity";
import FonjepTypePosteEntity from "../entities/FonjepTypePosteEntity";
import FonjepVersementEntity, { PayedFonjepVersementEntity } from "../entities/FonjepVersementEntity";

export const TIERS_ENTITY: FonjepTiersEntity = {
    code: "CodeAsso",
    raisonSociale: "Raison Sociale",
    estAssociation: "Oui",
    estCoFinanceurPostes: "Non",
    estFinanceurPostes: "Oui",
    siretOuRidet: DEFAULT_ASSOCIATION.siret,
    codePostal: "35700",
    ville: "Rennes",
    contactEmail: "john.doe@ille-et-vilaine.fr",
};

export const TIERS_ENTITY_WITH_NULLS: FonjepTiersEntity = {
    code: "CodeAsso2",
    siretOuRidet: null,
    raisonSociale: "Raison Sociale",
    estAssociation: "Oui",
    estCoFinanceurPostes: "Non",
    estFinanceurPostes: "Oui",
    codePostal: "35700",
    ville: "Rennes",
    contactEmail: "john.doe@ille-et-vilaine.fr",
};

export const POSTE_ENTITY: FonjepPosteEntity = {
    code: "Code",
    dispositifId: 1,
    financeurPrincipalCode: "FinanceurPrincipalCode",
    financeurAttributeurCode: "FinanceurAttributeurCode",
    associationBeneficiaireCode: TIERS_ENTITY.code,
    pstStatutPosteLibelle: "PstStatutPosteLibelle",
    pstRaisonStatutLibelle: "PstRaisonStatutLibelle",
    associationImplantationCode: "AssociationImplantationCode",
    annee: 2022,
    montantSubvention: 1000,
    dateFinTriennalite: new Date("2023-12-31"),
    pstTypePosteCode: "PstTypePosteCode",
    pleinTemps: "Oui",
    doublementUniteCompte: "Non",
};

export const POSTE_ENTITY_WITH_NULLS: FonjepPosteEntity = {
    code: "Code",
    dispositifId: null,
    financeurPrincipalCode: "FinanceurPrincipalCode",
    financeurAttributeurCode: "FinanceurAttributeurCode",
    associationBeneficiaireCode: TIERS_ENTITY_WITH_NULLS.code,
    pstStatutPosteLibelle: "PstStatutPosteLibelle",
    pstRaisonStatutLibelle: "PstRaisonStatutLibelle",
    associationImplantationCode: "AssociationImplantationCode",
    annee: 2022,
    montantSubvention: 1000,
    dateFinTriennalite: new Date("2023-12-31"),
    pstTypePosteCode: "PstTypePosteCode",
    pleinTemps: "Oui",
    doublementUniteCompte: "Non",
};

export const VERSEMENT_ENTITY: PayedFonjepVersementEntity = {
    posteCode: POSTE_ENTITY.code,
    periodeDebut: new Date("2022-01-12"),
    periodeFin: new Date("2022-12-14"),
    dateVersement: new Date("2022-04-15"),
    montantAPayer: 1000,
    montantPaye: 1000,
};

export const VERSEMENT_ENTITY_WITH_NULLS: FonjepVersementEntity = {
    posteCode: POSTE_ENTITY_WITH_NULLS.code,
    periodeDebut: null,
    periodeFin: null,
    dateVersement: null,
    montantAPayer: 1000,
    montantPaye: null,
};

export const TYPE_POSTE_ENTITY: FonjepTypePosteEntity = { code: "Code", libelle: "Libelle" };

export const DISPOSITIF_ENTITY: FonjepDispositifEntity = { id: 1, libelle: "Libelle", financeurCode: "FinanceurCode" };

export const TIERS_ENTITIES = [TIERS_ENTITY, TIERS_ENTITY_WITH_NULLS];

export const POSTE_ENTITIES = [POSTE_ENTITY, POSTE_ENTITY_WITH_NULLS];

export const VERSEMENT_ENTITIES = [VERSEMENT_ENTITY, VERSEMENT_ENTITY_WITH_NULLS];

export const TYPE_POSTE_ENTITIES = [TYPE_POSTE_ENTITY];

export const DISPOSITIF_ENTITIES = [DISPOSITIF_ENTITY];
