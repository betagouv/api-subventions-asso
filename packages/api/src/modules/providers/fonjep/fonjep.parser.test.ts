import FonjepSubventionEntity from './entities/FonjepSubventionEntity';
import FonjepVersementEntity from "./entities/FonjepVersementEntity";
import FonjepParser from './fonjep.parser';
import * as ParserHelper from "../../../shared/helpers/ParserHelper";
import { DATA_WITH_HEADER, DEFAULT_POSTE, DEFAULT_VERSEMENT } from "./__fixtures__/fonjepFileModels";
jest.mock('./entities/FonjepSubventionEntity');
jest.mock('./entities/FonjepVersementEntity');

describe("FonjepParser", () => {
    const indexDataByPathObjectMock = jest.spyOn(ParserHelper, "indexDataByPathObject");
    const PAGES = [
        [["foo", "bar"], ["foo1", "bar1"], ["foo2", "bar2"]],
        [["foo", "baz"], ["foo1", "baz1"], ["foo2", "baz2"]],
    ];

    // @ts-expect-error: mock
    (FonjepSubventionEntity as jest.Mock).mockImplementation((jest.fn()));
    // @ts-expect-error: mock
    (FonjepVersementEntity as jest.Mock).mockImplementation((jest.fn()));

    describe("mapHeaderToData()", () => {
        it("should return a mapped object with header as property name", () => {
            const expected = [[{ foo: "foo1", bar: "bar1" }, { foo: "foo2", bar: "bar2" }], [{ foo: "foo1", baz: "baz1" }, { foo: "foo2", baz: "baz2" }]];
            // @ts-expect-error: test private method
            const actual = FonjepParser.mapHeaderToData(PAGES);
            expect(actual).toEqual(expected);
        });
    });

    describe("findOnPropFactory()", () => {
        it("should return a function", () => {
            const expected = "function";
            // @ts-expect-error: test private method
            const actual = typeof FonjepParser.findOnPropFactory([], "propName");
            expect(actual).toEqual(expected);
        });
        it("should return a function that find given array on property", () => {
            const expected = { "Code": "5678" };
            const ARRAY = [{ "Code": "1234" }, expected];
            // @ts-expect-error: test private method
            const find = FonjepParser.findOnPropFactory(ARRAY, "Code");
            const actual = find(expected.Code);
            expect(actual).toEqual(expected);
        });
    });

    describe("filterOnPropFactory()", () => {
        it("should return a function", () => {
            const expected = "function";
            // @ts-expect-error: test private method
            const actual = typeof FonjepParser.filterOnPropFactory([], "propName");
            expect(actual).toEqual(expected);
        });
        it("should return a function that filter given array on property", () => {
            const CODE = "5678";
            const expected = [{ Code: CODE }, { Code: CODE }];
            const ARRAY = [{ "Code": "1234" }].concat(expected);
            // @ts-expect-error: test private method
            const filter = FonjepParser.filterOnPropFactory(ARRAY, "Code");
            const actual = filter(CODE);
            expect(actual).toEqual(expected);
        });
    });


    describe("createFonjepSubventionEntity()", () => {
        beforeAll(() => {
            indexDataByPathObjectMock.mockImplementation(jest.fn());
        });
        afterAll(() => {
            indexDataByPathObjectMock.mockReset();
        });
        it("should call ParserHelper to build indexed and legal informations", () => {
            // @ts-expect-error: test private method
            FonjepParser.createFonjepSubventionEntity({});
            expect(indexDataByPathObjectMock).toHaveBeenCalledTimes(2);
            expect(indexDataByPathObjectMock).toHaveBeenCalledWith(FonjepSubventionEntity.indexedProviderInformationsPath, {});
            expect(indexDataByPathObjectMock).toHaveBeenCalledWith(FonjepSubventionEntity.indexedLegalInformationsPath, {});
        })
        it("should create a FonjepSubventionEntity", () => {
            // @ts-expect-error: test private method;
            FonjepParser.createFonjepSubventionEntity({});
            const expected = 3;
            // @ts-expect-error: mock
            const actual = FonjepSubventionEntity.mock.calls[0].length;
            expect(actual).toEqual(expected);
        })
    });

    describe("createFonjepVersement()", () => {
        beforeAll(() => {
            indexDataByPathObjectMock.mockImplementation(jest.fn());
        });
        afterAll(() => {
            indexDataByPathObjectMock.mockReset();
        });

        it("should call ParserHelper to build indexed and legal informations", () => {
            // @ts-expect-error: test private method
            FonjepParser.createFonjepVersementEntity({});
            expect(indexDataByPathObjectMock).toHaveBeenCalledTimes(2);
            expect(indexDataByPathObjectMock).toHaveBeenCalledWith(FonjepVersementEntity.indexedProviderInformationsPath, {});
            expect(indexDataByPathObjectMock).toHaveBeenCalledWith(FonjepVersementEntity.indexedLegalInformationsPath, {});
        });

        it("should create a FonjepSubventionEntity", () => {
            // @ts-expect-error: test private method;
            FonjepParser.createFonjepVersementEntity({});
            const expected = 3;
            // @ts-expect-error: mock
            const actual = (FonjepVersementEntity as jest.Mock).mock.calls[0].length;
            expect(actual).toEqual(expected);
        });
    })

    describe("parse()", () => {
        const xlsParseMock = jest.spyOn(ParserHelper, "xlsParse");
        // @ts-expect-error: mock private method     
        const mapHeaderToDataMock = jest.spyOn(FonjepParser, "mapHeaderToData");
        // @ts-expect-error: mock private method     
        const createFonjepSubventionEntityMock = jest.spyOn(FonjepParser, "createFonjepSubventionEntity")
        // @ts-expect-error: mock private method     
        const createFonjepVersementEntityMock = jest.spyOn(FonjepParser, "createFonjepVersementEntity")

        beforeAll(() => {
            createFonjepSubventionEntityMock.mockImplementation(jest.fn());
            createFonjepVersementEntityMock.mockImplementation(jest.fn());
        })

        afterEach(() => {
            createFonjepSubventionEntityMock.mockClear();
            createFonjepVersementEntityMock.mockClear();
        })

        it("should call createFonjepSubventionEntity with parsedData", () => {
            xlsParseMock.mockImplementationOnce(jest.fn());
            mapHeaderToDataMock.mockImplementationOnce(() => DATA_WITH_HEADER);
            FonjepParser.parse({} as Buffer, new Date("2022-03-03"));
            // @ts-expect-error: test
            expect(createFonjepSubventionEntityMock.mock.calls[0][0]).toMatchSnapshot({ id: expect.any(String) });
        })
        it("should call createVersementFonjep with versement data", () => {
            xlsParseMock.mockImplementationOnce(jest.fn());
            mapHeaderToDataMock.mockImplementationOnce(() => DATA_WITH_HEADER);
            FonjepParser.parse({} as Buffer, new Date("2022-03-03"));
            expect(createFonjepVersementEntityMock).toHaveBeenCalledTimes(2);
            // @ts-expect-error: test
            expect(createFonjepVersementEntityMock.mock.lastCall).toMatchSnapshot([{ id: expect.any(String) }]);
        })
    });
});