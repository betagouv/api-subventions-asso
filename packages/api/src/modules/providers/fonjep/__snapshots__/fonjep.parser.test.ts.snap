// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FonjepParser parse() should call createFonjepEntity with parsedData 1`] = `
Object {
  "Annee": 2017,
  "Association": Object {
    "Code": 1234,
    "CodePostal": "00000",
    "ContactEmail": "exemple@beta.gouv.fr",
    "EstAssociation": "Oui",
    "EstCoFinanceurPostes": "Non",
    "EstFinanceurPostes": "Non",
    "RaisonSociale": "CENTRE D'INFO ****",
    "SiretOuRidet": "00000000000000",
    "Ville": "Paris",
  },
  "AssociationBeneficiaireCode": 1234,
  "AssociationImplantationCode": 1234,
  "Code": "J10540",
  "DateFinTriennalite": 43465,
  "DispositifId": 3,
  "DoublementUniteCompte": "Non",
  "Financeur": Object {
    "Code": 1234,
    "CodePostal": "00000",
    "ContactEmail": "exemple@beta.gouv.fr",
    "EstAssociation": "Oui",
    "EstCoFinanceurPostes": "Non",
    "EstFinanceurPostes": "Non",
    "RaisonSociale": "CENTRE D'INFO ****",
    "SiretOuRidet": "00000000000000",
    "Ville": "Paris",
  },
  "FinanceurAttributeurCode": 1234,
  "FinanceurPrincipalCode": "1234",
  "MontantSubvention": 6666,
  "PleinTemps": "Oui",
  "PstRaisonStatutLibelle": "Reconduction",
  "PstStatutPosteLibelle": "Attribu√©",
  "PstTypePosteCode": "FONJEP",
  "TypePoste": Object {
    "Code": "FONJEP",
    "Libelle": "Poste FONJEP",
  },
  "dispositif": Object {
    "FinanceurCode": 10008,
    "ID": 3,
    "Libelle": "FONJEP Politique de la ville",
  },
  "id": Any<String>,
  "updated_at": 2022-03-03T00:00:00.000Z,
}
`;
